### PR TITLE
Publish to Maven Central

### PR DESCRIPTION
**Pull request by pbielicki**
This PR aligns naming convention of Java packages and changes the publishing destination to Maven Central.

Naming alignment replaces `org.gradle.samples` package with `org.gradle.exemplar`. It also renames `sample-check` and `sample-discovery` modules to `check` and `discovery`. Finally the GAV is changed to from e.g. `org.gradle:sample-check:0.12.6` to `org.gradle.exemplar:check:0.12.6`.

Publishing is changed from `repo.gradle.org` to Maven Central.

The last change is adding TeamCity DSL to be able to run verification and publishing job in CI.

### CODE REVIEW

**Review by marcphilipp — APPROVED** [review]
Thanks! 👍

File: .teamcity/Project.kt [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    +:refs/heads/*
+""".trimIndent()
+        }
+    }
+
+    steps {
+        gradle {
+            useGradleWrapper = true
+            tasks = "check"
+            gradleParams = "--build-cache -Dgradle.cache.remote.username=%gradle.cache.remote.username% -Dgradle.cache.remote.password=%gradle.cache.remote.password%"
```
Comment (britter) [inline .teamcity/Project.kt:47]: ❌ Credentails must not be passed via cmd line arguments. We need to inject this via environment variables instead.

File: .teamcity/Project.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                githubUrl = "https://api.github.com"
+                authType = personalToken {
+                    token = "%github.bot-teamcity.token%"
+                }
+            }
+        }
+    }
+})
+
+object ReleasePlugin : BuildType({
```
Comment (britter) [inline .teamcity/Project.kt:65]: ⭕️ Exemplar is not a plugin but a library.

File: .teamcity/Project.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            buildFile = "build.gradle.kts"
+        }
+    }
+    params {
+        param("env.JAVA_HOME", "%linux.java8.oracle.64bit%")
+        param("env.GRADLE_CACHE_REMOTE_USERNAME", "%gradle.cache.remote.username%")
+        param("env.GRADLE_CACHE_REMOTE_PASSWORD", "%gradle.cache.remote.password%")
+        param("env.MAVEN_CENTRAL_STAGING_REPO_USER", "%mavenCentralStagingRepoUser%")
+        password("env.MAVEN_CENTRAL_STAGING_REPO_PASSWORD", "%mavenCentralStagingRepoPassword%")
+    }
```
Comment (britter) [inline .teamcity/Project.kt:96]: It looks to me like the GPG signing setup is missing. Artifacts that are supposed to be uploaded to Maven Central need to be signed.

File: build.gradle.kts [hunk 4]
```diff
@@ -1 +1,16 @@
+plugins {
+    id("io.github.gradle-nexus.publish-plugin") version "1.1.0"
+}
+
 version = "0.12.6"
```
Comment (britter) [inline build.gradle.kts:5]: Given that we changed the group and artifact IDs I think we can do a major version bump to 1.0.0.

File: .teamcity/Project.kt [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            useGradleWrapper = true
+            gradleParams = "--build-cache -Dgradle.publish.skip.namespace.check=true"
+            tasks = "publishMavenJavaPublicationToSonatypeRepository"
+            buildFile = "build.gradle.kts"
+        }
+    }
+    params {
+        param("env.JAVA_HOME", "%linux.java8.oracle.64bit%")
+        param("env.GRADLE_CACHE_REMOTE_USERNAME", "%gradle.cache.remote.username%")
+        param("env.GRADLE_CACHE_REMOTE_PASSWORD", "%gradle.cache.remote.password%")
```
Comment (britter) [inline .teamcity/Project.kt:93]:
```suggestion
        password("env.GRADLE_CACHE_REMOTE_PASSWORD", "%gradle.cache.remote.password%")
```

File: discovery/src/main/java/org/gradle/exemplar/loader/CommandsParser.java [hunk 6]
```diff
@@ -70,9 +70,9 @@ private static Command parseCommand(final Config commandConfig, final File sampl
         } catch (ConfigException e) {
             throw new InvalidSampleException("'executable' field cannot be empty", e);
         }
-        final String executionDirectory = ConfigUtil.string(commandConfig, EXECUTION_SUBDIRECTORY, null);
-        final List<String> commands = ConfigUtil.strings(commandConfig, ARGS, new ArrayList<String>());
-        final List<String> flags = ConfigUtil.strings(commandConfig, FLAGS, new ArrayList<String>());
+        final String executionDirectory = org.gradle.exemplar.loader.ConfigUtil.string(commandConfig, EXECUTION_SUBDIRECTORY, null);
+        final List<String> commands = org.gradle.exemplar.loader.ConfigUtil.strings(commandConfig, ARGS, new ArrayList<String>());
+        final List<String> flags = org.gradle.exemplar.loader.ConfigUtil.strings(commandConfig, FLAGS, new ArrayList<String>());
```
Comment (britter) [inline discovery/src/main/java/org/gradle/exemplar/loader/CommandsParser.java:75]:
```suggestion
        final String executionDirectory = ConfigUtil.string(commandConfig, EXECUTION_SUBDIRECTORY, null);
        final List<String> commands = ConfigUtil.strings(commandConfig, ARGS, new ArrayList<String>());
        final List<String> flags = ConfigUtil.strings(commandConfig, FLAGS, new ArrayList<String>());
```

File: discovery/src/main/java/org/gradle/exemplar/loader/CommandsParser.java [hunk 7]
```diff
@@ -84,9 +84,9 @@ private static Command parseCommand(final Config commandConfig, final File sampl
             }
         }
 
-        final boolean expectFailures = ConfigUtil.booleanValue(commandConfig, EXPECT_FAILURE, false);
-        final boolean allowAdditionalOutput = ConfigUtil.booleanValue(commandConfig, ALLOW_ADDITIONAL_OUTPUT, false);
-        final boolean allowDisorderedOutput = ConfigUtil.booleanValue(commandConfig, ALLOW_DISORDERED_OUTPUT, false);
+        final boolean expectFailures = org.gradle.exemplar.loader.ConfigUtil.booleanValue(commandConfig, EXPECT_FAILURE, false);
+        final boolean allowAdditionalOutput = org.gradle.exemplar.loader.ConfigUtil.booleanValue(commandConfig, ALLOW_ADDITIONAL_OUTPUT, false);
+        final boolean allowDisorderedOutput = org.gradle.exemplar.loader.ConfigUtil.booleanValue(commandConfig, ALLOW_DISORDERED_OUTPUT, false);
```
Comment (britter) [inline discovery/src/main/java/org/gradle/exemplar/loader/CommandsParser.java:89]:
```suggestion
        final boolean expectFailures = ConfigUtil.booleanValue(commandConfig, EXPECT_FAILURE, false);
        final boolean allowAdditionalOutput = ConfigUtil.booleanValue(commandConfig, ALLOW_ADDITIONAL_OUTPUT, false);
        final boolean allowDisorderedOutput = ConfigUtil.booleanValue(commandConfig, ALLOW_DISORDERED_OUTPUT, false);
```

File: docs/README.adoc [hunk 8]
```diff
@@ -10,7 +10,7 @@ toc::[]
 
 == Use cases
 
-The intent of sample-check is to ensure that users of your command-line tool see what you expect them to see.
+The intent of check is to ensure that users of your command-line tool see what you expect them to see.
```
Comment (britter) [inline docs/README.adoc:13]:
```suggestion
The intent of the `check` module is to ensure that users of your command-line tool see what you expect them to see.
```

**Review by britter — COMMENTED** [review]
Looking good, left a few comments

File: .teamcity/Project.kt (hunk 3 above)
Comment (pbielicki) [reply .teamcity/Project.kt]: done

File: .teamcity/Project.kt (hunk 1 above)
Comment (pbielicki) [reply .teamcity/Project.kt]: done

File: build.gradle.kts (hunk 4 above)
Comment (pbielicki) [reply build.gradle.kts]: done

File: buildSrc/src/main/kotlin/exemplar.publishing-conventions.gradle.kts [hunk 9]
```diff
@@ -1,26 +1,22 @@
...
-            }
-        }
-    }
+signing {
+    sign(publishing.publications["mavenJava"])
 }
+
+configure<SigningExtension> {
+    useInMemoryPgpKeys(System.getenv("PGP_SIGNING_KEY"), System.getenv("PGP_SIGNING_KEY_PASSPHRASE"))
+}
```
Comment (britter) [inline buildSrc/src/main/kotlin/exemplar.publishing-conventions.gradle.kts:22]: `configure<SigningExtension>` is the same as configuring via the `signing` block.

```suggestion
signing {
    sign(publishing.publications["mavenJava"])
    useInMemoryPgpKeys(System.getenv("PGP_SIGNING_KEY"), System.getenv("PGP_SIGNING_KEY_PASSPHRASE"))
}
```

File: settings.gradle.kts [hunk 10]
```diff
@@ -1,10 +1,12 @@
+import org.gradle.kotlin.dsl.support.serviceOf
+
```
Comment (britter) [inline settings.gradle.kts:2]: I don't think we need this, do we?

```suggestion
```

**Review by britter — APPROVED** [review]
LGTM, just two minor comments.

File: settings.gradle.kts (hunk 10 above)
Comment (pbielicki) [reply settings.gradle.kts]: blah - ugly leftover - nice catch 👍

File: buildSrc/src/main/kotlin/exemplar.publishing-conventions.gradle.kts (hunk 9 above)
Comment (pbielicki) [reply buildSrc/src/main/kotlin/exemplar.publishing-conventions.gradle.kts:22]: TIL 👍

**Pull request merged**